clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java

```diff
@@ -386,18 +386,39 @@ public static void writeDouble(double value, ByteBuffer buffer) {
         buffer.putDouble(value);
     }
 
+    final static int[] LEADING_ZEROS_TO_U_VARINT_SIZE = new int[] {
+        // 32 bits, and each 7-bits adds one byte to the output
+        5, 5, 5, 5, // 32
+        4, 4, 4, 4, 4, 4, 4, // 28
+        3, 3, 3, 3, 3, 3, 3, // 21
+        2, 2, 2, 2, 2, 2, 2, // 14
+        1, 1, 1, 1, 1, 1, 1, // 7
+        1 // 0
+    };
+
+    final static int[] LEADING_ZEROS_TO_U_VARLONG_SIZE = new int[] {
+        // 64 bits, and each 7-bits adds one byte to the output
+        10, // 64
+        9, 9, 9, 9, 9, 9, 9, // 63
+        8, 8, 8, 8, 8, 8, 8, // 56
+        7, 7, 7, 7, 7, 7, 7, // 49
+        6, 6, 6, 6, 6, 6, 6, // 42
+        5, 5, 5, 5, 5, 5, 5, // 35
+        4, 4, 4, 4, 4, 4, 4, // 28
+        3, 3, 3, 3, 3, 3, 3, // 21
+        2, 2, 2, 2, 2, 2, 2, // 14
+        1, 1, 1, 1, 1, 1, 1, // 7
+        1 // 0
+    };
+
     /**
      * Number of bytes needed to encode an integer in unsigned variable-length format.
      *
      * @param value The signed value
      */
     public static int sizeOfUnsignedVarint(int value) {
-        int bytes = 1;
-        while ((value & 0xffffff80) != 0L) {
-            bytes += 1;
-            value >>>= 7;
-        }
-        return bytes;
+        int leadingZeros = Integer.numberOfLeadingZeros(value);
+        return LEADING_ZEROS_TO_U_VARINT_SIZE[leadingZeros];
```

Comment:
Good suggestion, I added a benchmark in 87d4b19. Although, it looks like the div, even when compiled down to shift operations, is not as quick.
```
Benchmark                                             Mode  Cnt    Score   Error   Units
ByteUtilsBenchmark.testSizeOfUnsignedVarint          thrpt    5  290.900 ± 1.508  ops/ms
ByteUtilsBenchmark.testSizeOfUnsignedVarintMath      thrpt    5   70.522 ± 4.507  ops/ms
ByteUtilsBenchmark.testSizeOfUnsignedVarintOriginal  thrpt    5   87.415 ± 0.894  ops/ms
```

Looking at `-prof perfasm` it does use `lzcnt` and does compile down to bitshifts, they just turn out relatively expensive. This is a 5950X.

```
         │  0x00007f156148f435: movsxd  %edi,%r9
  0.00%  │  0x00007f156148f438: mov     0x14(%r8,%r9,4),%ecx
  7.66%  │  0x00007f156148f43d: lzcnt   %r11d,%r11d       ;*invokestatic numberOfLeadingZeros
         │                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMath@29 (line 63)
         │                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMath_jmhTest::testSizeOfUnsignedVarintMath_thrpt_jmhStub@17 (line 119)
  0.01%  │  0x00007f156148f442: mov     $0x26,%esi
         │  0x00007f156148f447: sub     %r11d,%esi        ;*isub
         │                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMath@39 (line 64)
         │                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMath_jmhTest::testSizeOfUnsignedVarintMath_thrpt_jmhStub@17 (line 119)
         │  0x00007f156148f44a: lzcnt   %ecx,%r9d         ;*invokestatic numberOfLeadingZeros
         │                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMath@29 (line 63)
         │                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMath_jmhTest::testSizeOfUnsignedVarintMath_thrpt_jmhStub@17 (line 119)
         │  0x00007f156148f44f: mov     $0x26,%ebp
         │  0x00007f156148f454: sub     %r9d,%ebp         ;*isub
         │                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMath@39 (line 64)
         │                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMath_jmhTest::testSizeOfUnsignedVarintMath_thrpt_jmhStub@17 (line 119)
         │  0x00007f156148f457: movsxd  %esi,%rcx
  0.01%  │  0x00007f156148f45a: movsxd  %ebp,%rbx
  7.57%  │  0x00007f156148f45d: imulq   $0x92492493,%rcx,%rcx
  0.00%  │  0x00007f156148f464: imulq   $0x92492493,%rbx,%rbx
         │  0x00007f156148f46b: sar     $0x20,%rcx
  0.01%  │  0x00007f156148f46f: sar     $0x20,%rbx
  0.01%  │  0x00007f156148f473: mov     %ecx,%ecx
         │  0x00007f156148f475: add     %esi,%ecx
  7.66%  │  0x00007f156148f477: mov     %ebx,%edx
  0.00%  │  0x00007f156148f479: add     %ebp,%edx
  0.00%  │  0x00007f156148f47b: sar     $0x2,%ecx
  7.59%  │  0x00007f156148f47e: sar     $0x2,%edx
  0.01%  │  0x00007f156148f481: movsxd  %ecx,%rbx
  7.71%  │  0x00007f156148f484: movsxd  %edx,%rdx
  0.00%  │  0x00007f156148f487: sar     $0x1f,%ebp
         │  0x00007f156148f48a: mov     %r9d,%ecx
         │  0x00007f156148f48d: sar     $0x1f,%ecx
         │  0x00007f156148f490: movsxd  %ebp,%rbp
         │  0x00007f156148f493: sub     %rbp,%rdx
  7.58%  │  0x00007f156148f496: shr     $0x1b,%ecx
         │  0x00007f156148f499: add     %r9d,%ecx
  0.00%  │  0x00007f156148f49c: sar     $0x1f,%esi
         │  0x00007f156148f49f: sar     $0x5,%ecx
         │  0x00007f156148f4a2: movsxd  %esi,%r9
         │  0x00007f156148f4a5: sub     %r9,%rbx
         │  0x00007f156148f4a8: movsxd  %ecx,%r9
  7.66%  │  0x00007f156148f4ab: add     %r9,%rdx
  7.66%  │  0x00007f156148f4ae: mov     %r11d,%r9d
         │  0x00007f156148f4b1: sar     $0x1f,%r9d
  0.00%  │  0x00007f156148f4b5: shr     $0x1b,%r9d
  0.00%  │  0x00007f156148f4b9: add     %r11d,%r9d
         │  0x00007f156148f4bc: sar     $0x5,%r9d
  7.73%  │  0x00007f156148f4c0: movsxd  %r9d,%r11
  7.83%  │  0x00007f156148f4c3: add     %rbx,%r11
  7.45%  │  0x00007f156148f4c6: add     %r13,%r11
  7.62%  │  0x00007f156148f4c9: add     %r11,%rdx         ;*ladd
         │                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMath@50 (line 64)
         │                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMath_jmhTest::testSizeOfUnsignedVarintMath_thrpt_jmhStub@17 (line 119)
```

Reply:
Yes, I agree - the arithmetic / bitshift looks intuitively quicker since it requires no loads. Looking at the compiler output from JDK11, it does get compiled down to a sequence of operations that are reasonably clean. Perhaps there is a slight advantage in different use-cases, but, in my case this fix takes the function completely out of the profiler results so it solves the problem.

Out of interest sake, here is the sequence of operations according to perfasm on JDK11.
```
         ││  0x00007f26e04797b7: mov     0x10(%r12,%r10,8),%r10d
  0.05%  ││  0x00007f26e04797bc: lzcnt   %r10d,%r11d       ;*invokestatic numberOfLeadingZeros {reexecute=0 rethrow=0 return_oop=0}
         ││                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMathOne@6 (line 76)
         ││                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMathOne_jmhTest::testSizeOfUnsignedVarintMathOne_thrpt_jmhStub@17 (line 119)
  0.03%  ││  0x00007f26e04797c1: mov     $0x26,%r8d
  0.01%  ││  0x00007f26e04797c7: sub     %r11d,%r8d        ;*isub {reexecute=0 rethrow=0 return_oop=0}
         ││                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMathOne@13 (line 77)
         ││                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMathOne_jmhTest::testSizeOfUnsignedVarintMathOne_thrpt_jmhStub@17 (line 119)
  0.49%  ││  0x00007f26e04797ca: mov     %r11d,%r9d
  0.36%  ││  0x00007f26e04797cd: sar     $0x1f,%r9d
  0.02%  ││  0x00007f26e04797d1: movsxd  %r8d,%r10
  0.77%  ││  0x00007f26e04797d4: shr     $0x1b,%r9d
  7.24%  ││  0x00007f26e04797d8: add     %r11d,%r9d
  0.78%  ││  0x00007f26e04797db: imulq   $0x92492493,%r10,%r10
  0.05%  ││  0x00007f26e04797e2: sar     $0x5,%r9d
  2.24%  ││  0x00007f26e04797e6: sar     $0x20,%r10
  9.93%  ││  0x00007f26e04797ea: movsxd  %r9d,%r11
  0.12%  ││  0x00007f26e04797ed: mov     %r10d,%r10d
         ││  0x00007f26e04797f0: add     %r8d,%r10d
  6.30%  ││  0x00007f26e04797f3: sar     $0x1f,%r8d
  2.53%  ││  0x00007f26e04797f7: sar     $0x2,%r10d
  6.21%  ││  0x00007f26e04797fb: movsxd  %r8d,%r8
         ││  0x00007f26e04797fe: movsxd  %r10d,%rdx
  7.25%  ││  0x00007f26e0479801: sub     %r8,%rdx
  7.01%  ││  0x00007f26e0479804: add     %r11,%rdx         ;*i2l {reexecute=0 rethrow=0 return_oop=0}
         ││                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintMathOne@22 (line 77)
         ││                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintMathOne_jmhTest::testSizeOfUnsignedVarintMathOne_thrpt_jmhStub@17 (line 119)
  8.11%  ││  0x00007f26e0479807: mov     (%rsp),%rsi
```

Lookup table:
```
         │  0x00007fde904792a7: mov     0x10(%r12,%r10,8),%r10d
  1.42%  │  0x00007fde904792ac: lzcnt   %r10d,%r10d       ;*invokestatic numberOfLeadingZeros {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - org.apache.kafka.common.utils.ByteUtils::sizeOfUnsignedVarint@1 (line 420)
         │                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintOne@6 (line 61)
         │                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintOne_jmhTest::testSizeOfUnsignedVarintOne_thrpt_jmhStub@17 (line 119)
  1.26%  │  0x00007fde904792b1: cmp     $0x21,%r10d
  2.65%  │  0x00007fde904792b5: jnb     0x7fde90479332
  0.02%  │  0x00007fde904792b7: mov     %r8,0x40(%rsp)
  0.02%  │  0x00007fde904792bc: movabs  $0xfe434bf0,%r11  ;   {oop([I{0x00000000fe434bf0})}
  1.68%  │  0x00007fde904792c6: movsxd  0x10(%r11,%r10,4),%rdx  ;*i2l {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - org.apache.kafka.jmh.util.ByteUtilsBenchmark::testSizeOfUnsignedVarintOne@9 (line 61)
         │                                                ; - org.apache.kafka.jmh.util.jmh_generated.ByteUtilsBenchmark_testSizeOfUnsignedVarintOne_jmhTest::testSizeOfUnsignedVarintOne_thrpt_jmhStub@17 (line 119)
 37.64%  │  0x00007fde904792cb: mov     (%rsp),%rsi
```

It may be that the code in yours is faster when dropped into a full application.

Perhaps there are a more direct series of shift operations that would be better suited?

Reply:
That does seem like a lot of instructions, some of them don't seem to be needed in this case where we deal with small unsigned integers (e.g. movsxd %r8d,%r8 is extending the sign to 64 bits, but we don't need it).  Java doesn't seem to support unsigned, so not sure if there is a way to hint the compiler that we don't need instructions that only matter for signed arithmetic.  Maybe using long would help at least to eliminate the need to do movsxd.

Reply:
I can get a decent bump in performance, almost to lookup table performance, by swapping the divide by 32 for a direct shift right. I'll need to work out the right incantation of casts to get a divide by 7 through shifts to give the correct answer.

Reply:
OK, got it. Thank you for the nudges. They are now effectively identical in performance, and the math-version uses 1 less cmp/mov (for array bounds & fetch), with one extra instruction, and no requirement to hit the cache.

```
Benchmark                                                                     Mode  Cnt       Score      Error      Units
ByteUtilsBenchmark.testSizeOfUnsignedVarint                                  thrpt   20  503633.665 ± 5214.083     ops/ms
ByteUtilsBenchmark.testSizeOfUnsignedVarint:IPC                              thrpt    2       4.608             insns/clk
ByteUtilsBenchmark.testSizeOfUnsignedVarint:L1-dcache-loads                  thrpt    2      13.002                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarint:branches                         thrpt    2       5.005                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarint:instructions                     thrpt    2      31.080                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarintMath                              thrpt   20  507438.241 ± 5268.695     ops/ms
ByteUtilsBenchmark.testSizeOfUnsignedVarintMath:CPI                          thrpt    2       0.209             clks/insn
ByteUtilsBenchmark.testSizeOfUnsignedVarintMath:L1-dcache-loads              thrpt    2      11.992                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarintMath:branches                     thrpt    2       4.004                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarintMath:instructions                 thrpt    2      32.059                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarintOriginal                          thrpt   20  371946.626 ± 3105.947     ops/ms
ByteUtilsBenchmark.testSizeOfUnsignedVarintOriginal:IPC                      thrpt    2       5.360             insns/clk
ByteUtilsBenchmark.testSizeOfUnsignedVarintOriginal:L1-dcache-loads          thrpt    2      14.002                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarintOriginal:branches                 thrpt    2       7.992                  #/op
ByteUtilsBenchmark.testSizeOfUnsignedVarintOriginal:instructions             thrpt    2      48.931                  #/op
```

Reply:
@artemlivshits i've updated the implementation in f040109ba.

Reply:
Nice!

Reply:
Sorry to chime in late here, but I'm wondering if a lookup table may still be faster. You can see these two instructions in your lookup table implementation above,
```
  1.26%  │  0x00007fde904792b1: cmp     $0x21,%r10d
  2.65%  │  0x00007fde904792b5: jnb     0x7fde90479332
```
this is comparing if the calculated number (leading zeroes) fits within the size-33 array. Does Java have the concept of a 256 sized array, and uint8 numbers?

In my Go implementation (which uses bits.Len, i.e., the opposite of leading zeroes), using a size-256 string and a uint8 allows for no bounds check, resulting in even fewer instructions and shaving a fraction of a nanosecond of time in comparison to the pure math version.

In this benchmark,
```go
func BenchmarkUvarintLen(b *testing.B) {
	z := 0
	for i := 0; i < b.N; i++ {
		z += UvarintLen(uint32(i))
	}
	b.Log(z)
}
```

Comparing the existing code to this code:
```go
func UvarintLen(u uint32) int {
        lz := bits.LeadingZeros32(u)
        return (((38 - lz) * 0b10010010010010011) >> 19) + (lz >> 5)}
```

Top is math version, bottom is lookup table:
```
BenchmarkUvarintLen-8   	1000000000	         1.179 ns/op
BenchmarkUvarintLen-8   	1000000000	         0.7811 ns/op
```

Reply:
I like your approach here, and it's something that _should_ be possible, but, I don't see it being applied, the below still includes the cmp/mov
```
final static byte[] LEADING_ZEROS_TO_U_VARINT_SIZE = new byte[4096];
```
Byte is -127 to +128 in Java, so 4096 should be plenty.
```
byte leadingZeros = (byte) Integer.numberOfLeadingZeros(inputInt);
return LEADING_ZEROS_TO_U_VARINT_SIZE[leadingZeros];
````

Reply:
Ah bummer, but interesting! Fun between languages :)